### DISCUSSION TITLE
Using glsl shaders with SMPlayer

### DISCUSSION THREAD

**Discussion by killpotts** [smplayer-dev/smplayer#346] [Q&A]
Hello there, I have been trying to use glsl shaders ( such as anime4k ) using mpv as the core without success. Searching all around the internet for information how how specifically this is done has not yielded good results.

I found this discussion: 

https://gitmetadata.com/repo/bloc97/Anime4K/issues/137

However, none of the syntax they seem to talk about actually produces any results other than crashing SMplayer upon playing any file with the following log error:
"Option vf-add: --glsl-shader doesn't exist."

Clearly, I am doing something wrong. Problem is there is no information about this error anywhere that I could find, google yields 0 results. There is, quite literally, zero documentation on how to do this in SMplayer. 

TL;DR I found info suggesting that this feat is possible in SMplayer, but no info on how exactly to do it.

**Comment by smplayer-dev** [comment]
You should add the ` --glsl-shader` option in Preferences -> Advanced -> mplayer/mpv -> Options, not in the Video filters field.

**Comment by killpotts** [comment]
Still doesn't work. I've been searching all over the place and cannot find an answer for this. 

The problem is, really, that there is just absolutely zero documentation on how to do it.  If this is supposed to be a feature, there needs to be some kind of syntax guide otherwise we are just guessing and stabbing in the dark until it works.

**Reply by tuqueque** [reply]
> Still doesn't work. I've been searching all over the place and cannot find an answer for this.
> 
> The problem is, really, that there is just absolutely zero documentation on how to do it. If this is supposed to be a feature, there needs to be some kind of syntax guide otherwise we are just guessing and stabbing in the dark until it works.

Hey, I hope my reply is helpful… I don't know if you're using Windows or Linux. I use Linux, so the exact details for windows will be slightly different. I'll try to point out those details where necessary.

1.- To use Anime4k shaders in mpv I imagine you followed the instructions to "install" Anime4K

in Windows:
https://github.com/bloc97/Anime4K/blob/master/md/GLSL_Instructions_Windows.md

In Linux:
https://github.com/bloc97/Anime4K/blob/master/md/GLSL_Instructions_Linux.md

An easy way of knowing if it's working, is simply opening a low-res video file (360p) directly in mpv (not SMPlayer), making the window bigger so the video automatically scales up to adapt to the bigger window... and playing with the Anime4K switches (Ctrl+1, Ctrl+2, Ctrl+3...). If you see the changes in the video, then things should be working correctly.

2.- Now, in SMPlayer, make sure you're using the correct "Output driver" and "Hardware decoding" options in the Preferences. In the case of my SMPlayer Linux install, is:
Options > Preferences > General > Video > Output driver: **gpu**
Options > Preferences > Performance > Hardware decoding: **vdpau**

I'm not sure the "vdpau" decoding option is available/correct in Windows. If you're in Linux, you might still need to experiment with this depending on your hardware (Nvidia, AMD, Intel...).

3.- Go to:
Options > Preferences > Advanced > MPlayer/mpv

In the "Options:" section, add the line to your shader(s)… In my case, I added the following:
```
--glsl-shaders="~/.config/mpv/shaders/Anime4K_Clamp_Highlights.glsl:~/.config/mpv/shaders/Anime4K_Restore_CNN_M.glsl:~/.config/mpv/shaders/Anime4K_Upscale_CNN_x2_M.glsl:~/.config/mpv/shaders/Anime4K_Restore_CNN_S.glsl:~/.config/mpv/shaders/Anime4K_AutoDownscalePre_x2.glsl:~/.config/mpv/shaders/Anime4K_AutoDownscalePre_x4.glsl:~/.config/mpv/shaders/Anime4K_Upscale_CNN_x2_S.glsl"
```

Be mindful of some things:
* The general command should be:
`--glsl-shaders="PATH_TO_YOUR_SHADERS_DIRECTORY\SHADER_NAME.glsl"`
* I added several shaders, they are separated by a colon (:) and the whole string of shaders are enclosed between double quotes (").
* In windows, you might need to add the whole path to your shaders, like this:
--glsl-shader=C:\Users\YOUR_USER\AppData\Roaming\mpv\shaders\adaptive-sharpen.glsl

So theoretically, if I convert my Anime4K command in SMPlayer to your Windows setup, it might look something like:
```
--glsl-shaders="C:\Users\YOUR_USER\AppData\Roaming\mpv\shaders\Anime4K_Clamp_Highlights.glsl:C:\Users\YOUR_USER\AppData\Roaming\mpv\shaders\Anime4K_Restore_CNN_M.glsl:C:\Users\YOUR_USER\AppData\Roaming\mpv\shaders\Anime4K_Upscale_CNN_x2_M.glsl:C:\Users\YOUR_USER\AppData\Roaming\mpv\shaders\Anime4K_Restore_CNN_S.glsl:C:\Users\YOUR_USER\AppData\Roaming\mpv\shaders\Anime4K_AutoDownscalePre_x2.glsl:C:\Users\YOUR_USER\AppData\Roaming\mpv\shaders\Anime4K_AutoDownscalePre_x4.glsl:C:\Users\YOUR_USER\AppData\Roaming\mpv\shaders\Anime4K_Upscale_CNN_x2_S.glsl"
```

Of course, change the "YOUR_USER" part in the path to your actual user in your system.

Hopefully this helps.
Greets.
